openmeter/streaming/clickhouse/cache.go

```diff
@@ -0,0 +1,276 @@
+	// If caching is disabled, we don't cache anything
+	if !c.config.QueryCacheEnabled {
+		return false
+	}
+
+	// We only cache queries where cachable is set to true
+	if !queryParams.Cachable {
+		return false
+	}
+
+	// If the namespace template is provided and does not match the namespace, we don't cache anything
+	if c.namespaceTemplate != nil && !c.namespaceTemplate.MatchString(namespace) {
+		return false
+	}
+
+	// We can only cache queries that have a from time
+	if queryParams.From == nil {
+		return false
+	}
+
+	from := *queryParams.From
+	to := lo.FromPtrOr(queryParams.To, time.Now().UTC())
+	duration := to.Sub(from)
+
+	// It must be at least 3 days of usage to be cachable
+	if duration < c.config.QueryCacheMinimumCacheableQueryPeriod {
+		return false
+	}
+
+	return meterDef.Aggregation == meterpkg.MeterAggregationSum || meterDef.Aggregation == meterpkg.MeterAggregationCount || meterDef.Aggregation == meterpkg.MeterAggregationMin || meterDef.Aggregation == meterpkg.MeterAggregationMax
+}
+
+// createRemainingQueryFactory returns a function that creates a new query meter starting from the end of the cached query meter
+// It is used to query the fresh rows for the remaining time period from the events table directly
+func (c *Connector) createRemainingQueryFactory(originalQueryMeter queryMeter) func(cachedQueryMeter queryMeter) queryMeter {
+	newQueryMeter := originalQueryMeter
+
+	return func(cachedQueryMeter queryMeter) queryMeter {
+		newQueryMeter := newQueryMeter
+
+		newQueryMeter.From = nil
+		newQueryMeter.FromExclusive = cachedQueryMeter.To
+
+		c.config.Logger.Debug("query fresh rows from events table", "from", newQueryMeter.From, "to", newQueryMeter.To)
+
+		return newQueryMeter
+	}
+}
+
+// executeQueryWithCaching queries the meter view and manages the cache, as:
+// 1. Look up cached rows
+// 2. Query new rows for the uncached time period
+// 3. Cache the new results
+// It returns the cached rows and the updated query meter.
+func (c *Connector) executeQueryWithCaching(ctx context.Context, hash string, originalQueryMeter queryMeter) (queryMeter, []meterpkg.MeterQueryRow, error) {
+	var values []meterpkg.MeterQueryRow
+
+	createRemainingQuery := c.createRemainingQueryFactory(originalQueryMeter)
+
+	// Calculate the period to query from the cache
+	cacheableQueryMeter, err := c.prepareCacheableQueryPeriod(originalQueryMeter)
+	if err != nil {
+		return originalQueryMeter, values, err
+	}
+
+	// Step 1: Look up cached rows
+	cachedValues, err := c.fetchCachedMeterRows(ctx, hash, cacheableQueryMeter)
+	if err != nil {
+		return originalQueryMeter, values, fmt.Errorf("failed to lookup cached meter rows: %w", err)
+	}
+
+	// If we have cached values, add them to the results
+	// Also, update the query range to query uncached periods
+	if len(cachedValues) > 0 {
+		c.config.Logger.Debug("cached rows found", "from", cacheableQueryMeter.From, "to", cacheableQueryMeter.To, "count", len(cachedValues))
```

Comment:
I would consider adding some metrics that would allow us to validate the hit rate of the cache, as frequent writes might make the cache entirely disabled.